pkg/ext/resources/tokens/store.go

```diff
@@ -0,0 +1,325 @@
+package tokens
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/pkg/auth/tokens/hashers"
+	"github.com/rancher/rancher/pkg/ext/resources/types"
+	v1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/v3/pkg/randomtoken"
+	authv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	authzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const TokenNamespace = "cattle-token-data"
+
+// +k8s:openapi-gen=false
+// +k8s:deepcopy-gen=false
+type TokenStore struct {
+	secretClient v1.SecretClient
+	secretCache  v1.SecretCache
+	sar          authzv1.SubjectAccessReviewInterface
+}
+
+func NewTokenStore(secretClient v1.SecretClient, secretCache v1.SecretCache, sar authzv1.SubjectAccessReviewInterface) types.Store[*RancherToken, *RancherTokenList] {
+	tokenStore := TokenStore{
+		secretClient: secretClient,
+		secretCache:  secretCache,
+		sar:          sar,
+	}
+	return &tokenStore
+}
+
+func (t *TokenStore) Create(ctx context.Context, userInfo user.Info, token *RancherToken, opts *metav1.CreateOptions) (*RancherToken, error) {
+	if token.Status.PlaintextToken == "" {
+		tokenValue, err := randomtoken.Generate()
+		if err != nil {
+			return nil, fmt.Errorf("unable to generate token value: %w", err)
+		}
+		token.Status.PlaintextToken = tokenValue
+		hashedValue, err := hashers.GetHasher().CreateHash(tokenValue)
+		if err != nil {
+			return nil, fmt.Errorf("unable to hash token value: %w", err)
+		}
+		token.Status.HashedToken = hashedValue
+	}
+	// user is trying to create a token for another user
+	if token.Spec.UserID != userInfo.GetName() {
+		authed, err := t.userHasFullPermissions(userInfo)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if user has * on tokens: %w", err)
+		}
+		if !authed {
+			return nil, fmt.Errorf("can't create token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
+		}
+	}
+	secret := secretFromToken(token)
+	_, err := t.secretClient.Create(secret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create secret for token: %w", err)
+	}
+	// users don't care about the hashed value
+	token.Status.HashedToken = ""
+	return token, nil
+}
+
+func (t *TokenStore) Update(ctx context.Context, userInfo user.Info, token *RancherToken, opts *metav1.UpdateOptions) (*RancherToken, error) {
+	// user is trying to update a token for another user
+	if token.Spec.UserID != userInfo.GetName() {
+		authed, err := t.userHasFullPermissions(userInfo)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if user has * on tokens: %w", err)
+		}
+		if !authed {
+			return nil, fmt.Errorf("can't create token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
```

Comment:
```suggestion
			return nil, fmt.Errorf("can't update token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
```